// @sentry/electron/main/integrations/index.js
const electronBreadcrumbs = require('./electron-breadcrumbs.js');
const onuncaughtexception = require('./onuncaughtexception.js');
const mainContext = require('./main-context.js');
const index = require('./sentry-minidump/index.js');
const electronMinidump = require('./electron-minidump.js');
const preloadInjection = require('./preload-injection.js');
const mainProcessSession = require('./main-process-session.js');
const browserWindowSession = require('./browser-window-session.js');
const additionalContext = require('./additional-context.js');
const netBreadcrumbs = require('./net-breadcrumbs.js');
const childProcess = require('./child-process.js');
const screenshots = require('./screenshots.js');
const rendererProfiling = require('./renderer-profiling.js');
const anr = require('./anr.js');

/* eslint-disable deprecation/deprecation */

exports.ElectronBreadcrumbs = electronBreadcrumbs.ElectronBreadcrumbs;
exports.OnUncaughtException = onuncaughtexception.OnUncaughtException;
exports.MainContext = mainContext.MainContext;
exports.SentryMinidump = index.SentryMinidump;
exports.ElectronMinidump = electronMinidump.ElectronMinidump;
exports.PreloadInjection = preloadInjection.PreloadInjection;
exports.MainProcessSession = mainProcessSession.MainProcessSession;
exports.BrowserWindowSession = browserWindowSession.BrowserWindowSession;
exports.AdditionalContext = additionalContext.AdditionalContext;
exports.Net = netBreadcrumbs.Net;
exports.ChildProcess = childProcess.ChildProcess;
exports.Screenshots = screenshots.Screenshots;
exports.RendererProfiling = rendererProfiling.RendererProfiling;
exports.Anr = anr.Anr;
