// @discordapp/discord-native-types/DiscordNative.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DesktopSources = exports.JSExceptionLocation = exports.StoredCrashInformation = void 0;
let StoredCrashInformation;
exports.StoredCrashInformation = StoredCrashInformation;

(function (StoredCrashInformation) {
  StoredCrashInformation[StoredCrashInformation["HasRTCConnection"] = 0] = "HasRTCConnection";
  StoredCrashInformation[StoredCrashInformation["IsSendingVideo"] = 1] = "IsSendingVideo";
  StoredCrashInformation[StoredCrashInformation["IsSendingStream"] = 2] = "IsSendingStream";
  StoredCrashInformation[StoredCrashInformation["IsReceivingVideo"] = 3] = "IsReceivingVideo";
  StoredCrashInformation[StoredCrashInformation["IsReceivingStream"] = 4] = "IsReceivingStream";
  StoredCrashInformation[StoredCrashInformation["VideoMediaSessionId"] = 5] = "VideoMediaSessionId";
  StoredCrashInformation[StoredCrashInformation["StreamMediaSessionId"] = 6] = "StreamMediaSessionId";
})(StoredCrashInformation || (exports.StoredCrashInformation = StoredCrashInformation = {}));

let JSExceptionLocation;
exports.JSExceptionLocation = JSExceptionLocation;

(function (JSExceptionLocation) {
  JSExceptionLocation[JSExceptionLocation["RendererProcessDelayed"] = 0] = "RendererProcessDelayed";
  JSExceptionLocation[JSExceptionLocation["RendererProcess"] = 1] = "RendererProcess";
  JSExceptionLocation[JSExceptionLocation["MainProcess"] = 2] = "MainProcess";
})(JSExceptionLocation || (exports.JSExceptionLocation = JSExceptionLocation = {}));

let DesktopSources;
exports.DesktopSources = DesktopSources;

(function (DesktopSources) {
  DesktopSources["WINDOW"] = "window";
  DesktopSources["SCREEN"] = "screen";
  DesktopSources["CAMERA"] = "camera";
})(DesktopSources || (exports.DesktopSources = DesktopSources = {}));
