// @discordapp/discord-native-types/DiscordNative.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StoredCrashInformation = exports.DesktopSources = void 0;
// For debug's sake, we JSON.stringify this. If a BigInt is ever added, it'll break that code, so
// be sure to adjust the debug print to not use JSON.stringify.
let StoredCrashInformation; // If this becomes widely expanded in use, consider including electron types as a dev dependency to pull in the complete type.
exports.StoredCrashInformation = StoredCrashInformation;
(function (StoredCrashInformation) {
  StoredCrashInformation[StoredCrashInformation["HasRTCConnection"] = 0] = "HasRTCConnection";
  StoredCrashInformation[StoredCrashInformation["IsSendingVideo"] = 1] = "IsSendingVideo";
  StoredCrashInformation[StoredCrashInformation["IsSendingStream"] = 2] = "IsSendingStream";
  StoredCrashInformation[StoredCrashInformation["IsReceivingVideo"] = 3] = "IsReceivingVideo";
  StoredCrashInformation[StoredCrashInformation["IsReceivingStream"] = 4] = "IsReceivingStream";
  StoredCrashInformation[StoredCrashInformation["VideoMediaSessionId"] = 5] = "VideoMediaSessionId";
  StoredCrashInformation[StoredCrashInformation["StreamMediaSessionId"] = 6] = "StreamMediaSessionId";
})(StoredCrashInformation || (exports.StoredCrashInformation = StoredCrashInformation = {}));
let DesktopSources;
exports.DesktopSources = DesktopSources;
(function (DesktopSources) {
  DesktopSources["WINDOW"] = "window";
  DesktopSources["SCREEN"] = "screen";
})(DesktopSources || (exports.DesktopSources = DesktopSources = {}));
