// @sentry/electron/renderer/index.js
const browser = require('@sentry/browser');
const index = require('./integrations/index.js');
const core = require('@sentry/core');
const sdk = require('./sdk.js');

const Integrations = Object.assign(Object.assign({}, index), browser.Integrations);

exports.BrowserClient = browser.BrowserClient;
exports.BrowserTracing = browser.BrowserTracing;
exports.Replay = browser.Replay;
exports.lastEventId = browser.lastEventId;
exports.showReportDialog = browser.showReportDialog;
exports.FunctionToString = core.FunctionToString;
exports.Hub = core.Hub;
exports.InboundFilters = core.InboundFilters;
exports.Scope = core.Scope;
exports.addBreadcrumb = core.addBreadcrumb;
exports.addGlobalEventProcessor = core.addGlobalEventProcessor;
exports.captureEvent = core.captureEvent;
exports.captureException = core.captureException;
exports.captureMessage = core.captureMessage;
exports.configureScope = core.configureScope;
exports.createTransport = core.createTransport;
exports.getCurrentHub = core.getCurrentHub;
exports.getHubFromCarrier = core.getHubFromCarrier;
exports.makeMain = core.makeMain;
exports.setContext = core.setContext;
exports.setExtra = core.setExtra;
exports.setExtras = core.setExtras;
exports.setTag = core.setTag;
exports.setTags = core.setTags;
exports.setUser = core.setUser;
exports.startTransaction = core.startTransaction;
exports.withScope = core.withScope;
exports.defaultIntegrations = sdk.defaultIntegrations;
exports.init = sdk.init;
exports.Integrations = Integrations;
